satellite/metainfo/metainfoloop.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package metainfo
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/internal/sync2"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/storage"
+)
+
+var (
+	// LoopError is a standard error class for this component
+	LoopError = errs.Class("metainfo loop error")
+)
+
+// Observer is an interface defining an observer that can subscribe to the metainfo loop
+type Observer interface {
+	RemoteSegment(context.Context, storj.Path, *pb.Pointer) error
+	RemoteObject(context.Context, storj.Path, *pb.Pointer) error
+	InlineSegment(context.Context, storj.Path, *pb.Pointer) error
+}
+
+// LoopConfig contains configurable values for the metainfo loop
+type LoopConfig struct {
+	CoalesceDuration time.Duration `help:"how long to wait for new observers before starting iteration" releaseDefault:"5s" devDefault:"5s"`
+	Interval         time.Duration `help:"how long to wait between metainfo iterations" releaseDefault:"30s" devDefault:"30s"`
+}
+
+// LoopService is a metainfo loop service
+type LoopService struct {
+	waitingObservers []Observer
+	observers        []Observer
+	Loop             *sync2.Cycle
```

Comment:
It shouldn't have a loop, the loop is triggered by the observers not anything else.